File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteRefresherCallback.java
package com.mapbox.services.android.navigation.v5.navigation;

import com.mapbox.api.directions.v5.models.DirectionsRoute;

import java.util.Date;

import timber.log.Timber;

class RouteRefresherCallback implements RefreshCallback {
  private final MapboxNavigation mapboxNavigation;
  private final RouteRefresher routeRefresher;

  RouteRefresherCallback(MapboxNavigation mapboxNavigation, RouteRefresher routeRefresher) {
    this.mapboxNavigation = mapboxNavigation;
    this.routeRefresher = routeRefresher;
  }

  @Override
  public void onRefresh(DirectionsRoute directionsRoute) {
    mapboxNavigation.startNavigation(directionsRoute, DirectionsRouteType.FRESH_ROUTE);
    routeRefresher.updateLastRefresh(new Date());
    routeRefresher.updateIsChecking(false);
  }

  @Override
  public void onError(RefreshError error) {
    Timber.w(error.getMessage());
    routeRefresher.updateIsChecking(false);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteRefresherCallback.kt
package com.mapbox.services.android.navigation.v5.internal.navigation

import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.services.android.navigation.v5.navigation.DirectionsRouteType
import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation
import com.mapbox.services.android.navigation.v5.navigation.RefreshCallback
import com.mapbox.services.android.navigation.v5.navigation.RefreshError
import java.util.Date
import timber.log.Timber

internal class RouteRefresherCallback(
    private val mapboxNavigation: MapboxNavigation,
    private val routeRefresher: RouteRefresher
) : RefreshCallback {

    override fun onRefresh(directionsRoute: DirectionsRoute) {
        mapboxNavigation.startNavigation(directionsRoute, DirectionsRouteType.FRESH_ROUTE)
        routeRefresher.updateLastRefresh(Date())
        routeRefresher.updateIsChecking(false)
    }

    override fun onError(error: RefreshError) {
        error.message?.let {
            Timber.w(it)
        }
        routeRefresher.updateIsChecking(false)
    }
}
